provider/datagen/src/baked_exporter.rs

```diff
@@ -379,290 +319,298 @@ impl DataExporter for BakedExporter {
-                    #(#any_cases)*
-                    _ => return Err(DataErrorKind::MissingDataKey.with_req(key, req)),
+                    #(
+                        #features
+                        #hash_idents => #lookups #into_any_payloads,
+                    )*
+                    _ => return Err(icu_provider::DataErrorKind::MissingDataKey.with_req(key, req)),
                 }
-                .map(|payload| AnyResponse {
+                .map(|payload| icu_provider::AnyResponse {
                     payload: Some(payload),
                     metadata: Default::default(),
                 })
-                .ok_or_else(|| DataErrorKind::MissingLocale.with_req(key, req))
+                .ok_or_else(|| icu_provider::DataErrorKind::MissingLocale.with_req(key, req))
             }
         };
 
-        let mods = self
-            .mod_files
-            .get_mut()
-            .expect("poison")
-            .remove(&PathBuf::new())
-            .unwrap_or_default()
-            .into_iter()
-            .map(|p| p.parse::<TokenStream>().unwrap());
-
-        let data_impls = data_impls.values();
-
         self.write_to_file(
-            PathBuf::from("mod"),
+            PathBuf::from("macros"),
             quote! {
-                #(
-                    #[clippy::msrv = "1.61"]
-                    mod #mods;
-                )*
-
-                #[clippy::msrv = "1.61"]
-                use icu_provider::prelude::*;
-
-                /// Implement [`DataProvider<M>`] on the given struct using the data
-                /// hardcoded in this module. This allows the struct to be used with
+                /// Implement [`DataProvider<M>`](icu_provider::DataProvider) on the given struct using the data
+                /// hardcoded in this file. This allows the struct to be used with
                 /// `icu`'s `_unstable` constructors.
                 ///
-                /// This macro can only be called from its definition-site, i.e. right
-                /// after `include!`-ing the generated module.
-                ///
                 /// ```compile_fail
                 /// struct MyDataProvider;
-                /// include!("/path/to/generated/mod.rs");
+                /// include!("/path/to/generated/macros.rs");
                 /// impl_data_provider(MyDataProvider);
                 /// ```
-                #[allow(unused_macros)]
-                macro_rules! impl_data_provider {
+                #[doc(hidden)]
+                #[macro_export]
+                macro_rules! __impl_data_provider {
                     ($provider:path) => {
-                        #(#data_impls)*
+                        #(
+                            #features
+                            #macro_idents ! ($provider);
+                        )*
                     }
                 }
+                #[doc(inline)]
+                pub use __impl_data_provider as impl_data_provider;
```

Comment:
Question: why the rename and re-export?

Reply:
Because macros are weird. `macro_export` will put them at the root of the crate, so we hide that, and `pub use` respects module structure. This is explained in a comment in line 465.